src/datachain/utils.py

```diff
@@ -225,30 +238,65 @@ def get_envs_by_prefix(prefix: str) -> dict[str, str]:
 _T_co = TypeVar("_T_co", covariant=True)
 
 
-def batched(iterable: Iterable[_T_co], n: int) -> Iterator[tuple[_T_co, ...]]:
-    """Batch data into tuples of length n. The last batch may be shorter."""
-    # Based on: https://docs.python.org/3/library/itertools.html#itertools-recipes
-    # batched('ABCDEFG', 3) --> ABC DEF G
-    if n < 1:
-        raise ValueError("Batch size must be at least one")
-    it = iter(iterable)
-    while batch := tuple(islice(it, n)):
+def _dynamic_batched_core(
+    iterable: Iterable[_T_co],
+    chunk_rows: int,
+    chunk_mb: float,
+) -> Iterator[list[_T_co]]:
+    """Core batching logic that yields lists."""
+    max_memory_bytes = chunk_mb * 1024 * 1024
+
+    batch: list[_T_co] = []
+    current_memory = 0
+
+    for row_count, item in enumerate(iterable):
+        item_memory = estimate_memory_recursive(item)
```

Comment:
We cannot afford doing this for every item - too expensive. The logic need to be changes.